Find ethereum and hedera transactions by eth hash

dimitrovmaksim opened the pull request:
**Description**:

Allows fetching Ethereum and Hedera transaction by eth hash using the `contract_result` table as discussed here https://github.com/hashgraph/hedera-mirror-node/pull/4176. Opened a new PR to start on a clean base.

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-mirror-node/issues/4175

**Notes for reviewer**:
Couple potential Issues:

1. Getting the transaction by `ID` still requires fetching the transaction from the `transaction` table, because `contract_result` does not have `valid_start_ns` column, which is required for that query. This results in queries by ID returning the full hedera hash, instead of the short eth hash, compared to fetching the transaction by hash, which will always return the short hash.
2. Fetching the transaction by `contract_id` and `consensus_timestamp` still uses the transaction table query. Looking into the tests i'm not sure if making that query use the `contract_result` table won't cause issues ([example test](https://github.com/hashgraph/hedera-mirror-node/blob/main/hedera-mirror-rest/__tests__/specs/contracts/%7Bid%7D/results/%7Bid%7D/timestamp-result-not-found.json))
3. Also it seems that `contract_id` is completely ignored when querying transactions by contract_id/address and timestamp, and the contract_id/address are just format validated. More info here: https://github.com/hashgraph/hedera-mirror-node/issues/4380


 
**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

dimitrovmaksim commented:
> LGTM. not sure why some of the tests are failing

The chart check fails because it tries to fetch https://raw.githubusercontent.com/**_hashgraph_**/hedera-mirror-node/4176-find-eth-and-hedera-tx-by-hash/hedera-mirror-rosetta/scripts/validation/postman/rosetta-api-postman.json ,but the correct url is https://raw.githubusercontent.com/**_dimitrovmaksim_**/hedera-mirror-node/4176-find-eth-and-hedera-tx-by-hash/hedera-mirror-rosetta/scripts/validation/postman/rosetta-api-postman.json

steven-sheehy commented:
> Not sure if this is because I'm opening the PR from a fork (I don't think i can push directly to the repo)

Right, we've been aware of it but hadn't gotten around to addressing it. There's a PR up now to fix it: https://github.com/hashgraph/hedera-mirror-node/pull/4387

dimitrovmaksim commented:
> Right, we've been aware of it but hadn't gotten around to addressing it. There's a PR up now to fix it: #4387

Looks like it's working now 👍🏻

dimitrovmaksim commented:
Hm, now that we can find non-ethereum transactions by the eth hash, we should somehow return nonce for the non-eth transactions as well

Nana-EC commented on hedera-mirror-rest/__tests__/specs/contracts/results/{id}/by-hash-no-eth-tx.json:
```diff
@@ -0,0 +1,153 @@
+        "consensus_timestamp": 167654000123456,
+        "payerAccountId": "8001",
+        "transaction_hash": "4a563af33c4871b51a8b108aa2fe1dd5280a30dfb7236170ae5e5e7957eb639200000000000000000000000000000000",
+        "type": 50,
```
Type shouldn't be `50: 'ETHEREUMTRANSACTION'`
Try `7: 'CONTRACTCALL'` or ` 8: 'CONTRACTCREATEINSTANCE'`

Nana-EC reviewed (changes requested):
Looks good.
Just need to fix transaction type in test for non `ETHEREUMTRANSACTION` case

steven-sheehy commented:
> Fetching the transaction by contract_id and consensus_timestamp still uses the transaction table query. Looking into the tests i'm not sure if making that query use the contract_result table won't cause issues ([example test](https://github.com/hashgraph/hedera-mirror-node/blob/main/hedera-mirror-rest/__tests__/specs/contracts/%7Bid%7D/results/%7Bid%7D/timestamp-result-not-found.json))

I do think we should also make `/api/v1/contracts/{id}/results/{timestamp}` and ``/api/v1/contracts/results/{timestamp}` not use the transaction table. I'm not sure the referenced test is valid anymore since we now create contract_result even for failures after https://github.com/hashgraph/hedera-mirror-node/pull/3182

dimitrovmaksim commented on hedera-mirror-rest/service/transactionService.js:
```diff
@@ -129,7 +141,7 @@ class TransactionService extends BaseService {
         `${EthereumTransaction.PAYER_ACCOUNT_ID} = $1`
```
Well, then doesn't this also apply for the timestamp query?

dimitrovmaksim commented:
Hey @steven-sheehy @Nana-EC I've pushed some changes, hopefully I understood and addressed the comments correctly.
Also I think this comments still remains as an issue https://github.com/hashgraph/hedera-mirror-node/pull/4381#issuecomment-1233061413

steven-sheehy reviewed (changes requested):
Looking good, just some nit picks and need to fix merge conflicts

dimitrovmaksim commented on hedera-mirror-rest/service/contractService.js:
```diff
@@ -40,6 +48,38 @@ class ContractService extends BaseService {
                                                                     from ${ContractResult.tableName} ${ContractResult.tableAlias}
   `;
 
+  static detailedContractResultsWithEthereumTransactionQuery = `select ${ContractResult.tableAlias}.*,
```
Hm, strange, maybe I have triggered a linter, because I'm pretty sure I did not format it like this 🤔

steven-sheehy replied:
Right, a linter does it automatically. But I believe if you put a newline after the ` you can force it to start earlier on the line.

dimitrovmaksim commented on hedera-mirror-rest/service/contractService.js:
```diff
@@ -40,6 +48,38 @@ class ContractService extends BaseService {
+                                                                    ContractResult.TRANSACTION_HASH
+                                                                  )} as hash,
+                                                                  ${EthereumTransaction.tableAlias}.*,
+                                                                coalesce(
+                                                                  ${EthereumTransaction.getFullName(
+                                                                    EthereumTransaction.CONSENSUS_TIMESTAMP
+                                                                  )},
+                                                                  ${ContractResult.getFullName(
+                                                                    ContractResult.CONSENSUS_TIMESTAMP
+                                                                  )}
+                                                                ) as ${ContractResult.CONSENSUS_TIMESTAMP},
```
Hm i think I'll have to explicitly select both columns from the ContractResult table and set an alias, otherwise I think the columns will be prefixed with the tables aliases, because both columns exist in both tables.

steven-sheehy replied:
Why? The consensus timestamp is a join column so it will be the same value in each column. So you just need to show one, no coalescing.

dimitrovmaksim replied:
What i meant is that i'll have to select consensus_timestamp and gas_limit from the contract_result table explicitly, but i think re-arranging the query also fixed it

dimitrovmaksim commented on hedera-mirror-rest/service/contractService.js:
```diff
@@ -192,10 +256,59 @@ class ContractService extends BaseService {
     const rows = await super.getRows(query, params, 'getContractResultsByTimestamps');
-    return rows.map((row) => new ContractResult(row));
+
+    return [rows.map((row) => new TransactionWithEthData(row)), rows.map((row) => new ContractResult(row))];
```
Technically the data represents a transaction, even if it's not fetched from the transaction table, also wouldn't it be more misleading if the contract result is not actually for an ethereum transaction, but is called EthereumTransaction?

dimitrovmaksim replied:
I think I get the idea, and if we want to make this change and EthereumTransaction to represent only ethereum transactions, and non-eth transactions to be represented by contractResult, we have to probably make changes to `ContractResultDetailsViewModel` to fetch some of the shared columns from contractResult, instead of transaction.

steven-sheehy commented on hedera-mirror-rest/controllers/contractController.js:
```diff
@@ -976,56 +974,65 @@ class ContractController extends BaseController {
+      }
+
+      // Fetch transactions details and contractResults by mapped timestamp
+      [ethTransactions, contractResults] = await ContractService.getContractResultsByTimestamps(
```
We should still do this operation in parallel with logs, state changes and block retrieval. We don't need to use `ethTransaction.consensusTimestamp` and can use the `transaction.consensusTimestamp` for all.

dimitrovmaksim replied:
This will work when fetching the transaction by Id, but what about when we try to fetch the transaction by hash?

steven-sheehy replied:
Actually, this highlights the fact that we're doing duplicate calls to get contract results for search by hash, right? Once for lookup by hash then again for lookup by timestamp. So you're right that we can't do in parallel for search by hash but we should fix the duplicate calls.

dimitrovmaksim replied:
That was the reason I moved the ContractResult queries to not be done in parallel, because previously it was doing duplicate calls. Currently for hash it does the same query as for timestamp, but it's filtering by hash instead.

steven-sheehy replied:
Nevermind, with the diff and the large else statement I couldn't tell that it wasn't duplicated. It looks good to me now on a second look.

dimitrovmaksim commented on hedera-mirror-rest/controllers/contractController.js:
```diff
@@ -976,56 +974,65 @@ class ContractController extends BaseController {
-      RecordFileService.getRecordFileBlockDetailsFromTimestamp(transaction.consensusTimestamp),
-      ContractService.getContractLogsByTimestamps(transaction.consensusTimestamp),
-      ContractService.getContractStateChangesByTimestamps(transaction.consensusTimestamp),
+    if (contractResults.length === 0) {
+      throw new NotFoundError();
+    }
```
We should at least check that the consensusTimestamp is not undefined before doing the RecordFile, ContractLog and StateChange queries, maybe move the other check here and use the contractResult's consesusTimestamp for those queries?

steven-sheehy replied:
If there is at least one contract result row or transaction row then the consensus timestamp within those is guaranteed to be present since it's a primary key of those.

dimitrovmaksim commented on hedera-mirror-rest/controllers/contractController.js:
```diff
@@ -976,56 +974,65 @@ class ContractController extends BaseController {
       // should always return a contract results when
       // contract results are empty AND transaction type = ethereum transaction
-      isFailedContractResult = transaction.transactionType.toString() === ethereumTransactionType;
+      isFailedContractResult = ethTransaction.transactionType.toString() === ethereumTransactionType;
```
Is it even possible now to have an EthereumTransaction without a ContractResult?

steven-sheehy replied:
It is still possible. I believe we always create a result for contract call and contract create, but not for EthereumTransaction. It can occur when the ethereum transaction has an invalid payload and doesn't invoke the EVM.

dimitrovmaksim replied:
We currently do a left join, which if I'm not wrong won't return a row if a ContractResult with a specified hash/timestamp does not exists, but there's a matching record in the EthereumTransaction table

steven-sheehy replied:
If you think you see a bug, let's open a ticket. Meanwhile can we address the feedback so we can merge this?

dimitrovmaksim commented on hedera-mirror-rest/service/contractService.js:
```diff
@@ -224,10 +240,58 @@ class ContractService extends BaseService {
+    }
+
+    const whereClause = `where ${ContractResult.getFullName(ContractResult.TRANSACTION_HASH)} = $1`;
+    const query = [
+      ContractService.ethereumTransactionTableFullCTE,
+      ContractService.detailedContractResultsWithEthereumTransactionQuery,
+      ContractService.joinEthereumTransactionTable,
+      whereClause,
+      transactionsFilter,
+      super.getOrderByQuery(OrderSpec.from(ContractResult.getFullName(ContractResult.CONSENSUS_TIMESTAMP), 'asc')),
+      limit ? `limit ${limit}` : '',
+    ].join('\n');
```
Yeah, we discussed this here https://github.com/hashgraph/hedera-mirror-node/pull/4381#discussion_r967597201.

If I understand correctly:
1. If we leave it as it is, we will introduce a bug with this PR.
2. If we separate the queries, we will have once again to join `ethereum_transaction` and `transaction` in order to filter on the `transaction.result` until this `we can consider make another importer change to also persist transaction.result in ethereum_transaction table.` is eventually resolved
3. Can we leave it as it is atm, but if `contractResults` array is empty, we do an additional `ethereum_transation` query filtering on the hash/timestamp and see if there’s an `ethereum_transaction`, and if there’s one, create a default contract result, again until `result` is brought to the `ethereum_transaction` table and the queries can be separated?

Also one more thing worth bringing attention to, now that we query the `contract_result` table, we return `null` as `nonce` for the regular transactions, which is a problem. I can temporarily set it to return 0, instead of `null`, if this is okay

cc: @steven-sheehy @xin-hedera

xin-hedera replied:
approach 3 won't work, because with `contract_result cr left join etherem_transaciton ... where cr.hash = $1`, for when f ethereum transactions fail with certain codes, there won't be a corresponding row in `contract_result`.

Then for the rest API query with a transaction hash resulting in no rows returned from the left join, we can't tell if it's one of the two:

- No transaction (either normal SC or ethereum transaction) exists with the queried hash
- An ethereum transaction with those no-contract-result-row failures

dimitrovmaksim replied:
So approach 2? Move back the `ethereum_transaction` join on `transaction` to the TransactionService, query transactions and contract_results in parallel then check if contract results exists?

xin-hedera replied:
we can do that, and we can remove the join between `ethereum_transaction` and `transaction` when we make the importer change.
